File: apps/agent-worker/worker/scenario_engine/skillarts_strategy.py
"""SkillArts tiered PRB cadence (parity with skillarts_strategy.ts)."""

from __future__ import annotations

from typing import Any, Literal, Mapping, TypedDict

from worker.scenario_engine.date_calendar import add_calendar_days_utc, parse_iso_date_utc
from worker.scenario_engine.prb_weekday_anchors import (
    first_working_thursday_on_or_after,
    resolve_prb_anchor_day,
)
from worker.scenario_engine.scenario_linear_shared import (
    linear_allow_non_working_for,
    linear_clamp_days,
    linear_prb_step_allow_non_working,
    linear_step_payload,
)
from worker.scenario_engine.timing_profiles import uses_skillarts_tiered_prb_cadence
from worker.scenario_engine.working_days import (
    HolidaySet,
    add_working_days_utc,
    inclusive_working_day_span,
    next_working_day,
    previous_working_day_on_or_before,
)

SA_PRB_IDS = frozenset({
    'submit_prb1',
    'prb1_review',
    'submit_prb2',
    'prb2_review',
    'submit_prb3',
    'prb3_review',
})


class SkillArtsPrbRefs(TypedDict):
    skill_arts_prb1_submit_start: str | None
    skill_arts_prb1_submit_end: str | None
    skill_arts_prb2_submit_start: str | None
    skill_arts_prb2_submit_end: str | None
    skill_arts_prb3_submit_start: str | None
    skill_arts_prb3_submit_end: str | None


def skill_arts_apply_pinned_prefix_prb_state(
    timing_profile: str,
    allow: bool,
    row: Mapping[str, Any],
    pin: Mapping[str, Any],
    refs: SkillArtsPrbRefs,
) -> None:
    rid = str(row['id'])
    if not uses_skillarts_tiered_prb_cadence(timing_profile) or allow or rid not in SA_PRB_IDS:
        return
    if rid == 'submit_prb1':
        refs['skill_arts_prb1_submit_start'] = str(pin['start_date'])
        refs['skill_arts_prb1_submit_end'] = str(pin['end_date'])
    elif rid == 'submit_prb2':
        refs['skill_arts_prb2_submit_start'] = str(pin['start_date'])
        refs['skill_arts_prb2_submit_end'] = str(pin['end_date'])
    elif rid == 'submit_prb3':
        refs['skill_arts_prb3_submit_start'] = str(pin['start_date'])
        refs['skill_arts_prb3_submit_end'] = str(pin['end_date'])


def skill_arts_adjust_effective_for_development_dev_gap(
    timing_profile: str,
    allow: bool,
    row: Mapping[str, Any],
    holidays: HolidaySet,
    allow_map: dict[str, bool] | None,
    mod_sum: int,
    effective: int,
    skill_arts_tier_span: int,
    refs: SkillArtsPrbRefs,
) -> int | str:
    rid = str(row['id'])
    dev_prb_row = rid in ('development_prb1', 'development_prb2', 'development_prb3')
    use_skillarts_tiered_dev_gap = uses_skillarts_tiered_prb_cadence(timing_profile) and not allow and dev_prb_row
    if not use_skillarts_tiered_dev_gap:
        return effective

    submit_anchor_a: str | None = None
    submit_end_sa: str | None = None
    review_phase_s: Literal['prb1_review', 'prb2_review', 'prb3_review'] = 'prb1_review'
    err_early_s = 'PRB1 development scheduled before PRB1 submit (internal error).'
    if rid == 'development_prb1':
        submit_anchor_a = refs['skill_arts_prb1_submit_start']
        submit_end_sa = refs['skill_arts_prb1_submit_end']
        review_phase_s = 'prb1_review'
        err_early_s = 'PRB1 development scheduled before PRB1 submit (internal error).'
    elif rid == 'development_prb2':
        submit_anchor_a = refs['skill_arts_prb2_submit_start']
        submit_end_sa = refs['skill_arts_prb2_submit_end']
        review_phase_s = 'prb2_review'
        err_early_s = 'PRB2 development scheduled before PRB2 submit (internal error).'
    else:
        submit_anchor_a = refs['skill_arts_prb3_submit_start']
        submit_end_sa = refs['skill_arts_prb3_submit_end']
        review_phase_s = 'prb3_review'
        err_early_s = 'PRB3 development scheduled before PRB3 submit (internal error).'
    if not submit_anchor_a or not submit_end_sa:
        return err_early_s
    ideal_review = add_working_days_utc(submit_anchor_a, skill_arts_tier_span - 1, holidays)
    review_allow_s = linear_allow_non_working_for(review_phase_s, allow_map)
    review_resolved_s = resolve_prb_anchor_day(ideal_review, holidays, review_allow_s)
    gap_start = next_working_day(add_calendar_days_utc(submit_end_sa, 1), holidays)
    gap_end = previous_working_day_on_or_before(
        add_calendar_days_utc(str(review_resolved_s['iso']), -1), holidays
    )
    if parse_iso_date_utc(gap_start) > parse_iso_date_utc(gap_end):
        gap_start = gap_end
    raw_span = max(1, inclusive_working_day_span(gap_start, gap_end, holidays))
    bumped = linear_clamp_days(raw_span + mod_sum, row)
    return min(bumped, raw_span)


def skill_arts_try_place_tiered_prb_row(
    timing_profile: str,
    allow: bool,
    row: Mapping[str, Any],
    cursor_cal: str,
    holidays: HolidaySet,
    allow_map: dict[str, bool] | None,
    effective: int,
    note_out: str,
    skill_arts_tier_span: int,
    refs: SkillArtsPrbRefs,
    steps_out: list[dict[str, Any]],
) -> tuple[Literal['placed'], str] | tuple[Literal['skip'], None] | tuple[Literal['err'], str]:
    rid = str(row['id'])
    use_skillarts_tiered_cadence = (
        uses_skillarts_tiered_prb_cadence(timing_profile) and not allow and rid in SA_PRB_IDS
    )
    if not use_skillarts_tiered_cadence:
        return ('skip', None)

    phase_allow = linear_allow_non_working_for(rid, allow_map)

    if rid == 'submit_prb1':
        min_start = next_working_day(cursor_cal, holidays)
        merged = min_start
        thu_raw = first_working_thursday_on_or_after(merged, holidays)
        resolved = resolve_prb_anchor_day(thu_raw, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        refs['skill_arts_prb1_submit_start'] = start
        refs['skill_arts_prb1_submit_end'] = end
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    if rid == 'prb1_review':
        if not refs['skill_arts_prb1_submit_start']:
            return ('err', 'PRB1 review scheduled before PRB1 submit (internal error).')
        ideal_review = add_working_days_utc(refs['skill_arts_prb1_submit_start'], skill_arts_tier_span - 1, holidays)
        resolved = resolve_prb_anchor_day(ideal_review, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    if rid == 'submit_prb2':
        min_start = next_working_day(cursor_cal, holidays)
        merged = min_start
        thu_raw = first_working_thursday_on_or_after(merged, holidays)
        resolved = resolve_prb_anchor_day(thu_raw, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        refs['skill_arts_prb2_submit_start'] = start
        refs['skill_arts_prb2_submit_end'] = end
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    if rid == 'prb2_review':
        if not refs['skill_arts_prb2_submit_start']:
            return ('err', 'PRB2 review scheduled before PRB2 submit (internal error).')
        ideal_review = add_working_days_utc(refs['skill_arts_prb2_submit_start'], skill_arts_tier_span - 1, holidays)
        resolved = resolve_prb_anchor_day(ideal_review, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    if rid == 'submit_prb3':
        min_start = next_working_day(cursor_cal, holidays)
        merged = min_start
        thu_raw = first_working_thursday_on_or_after(merged, holidays)
        resolved = resolve_prb_anchor_day(thu_raw, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        refs['skill_arts_prb3_submit_start'] = start
        refs['skill_arts_prb3_submit_end'] = end
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    if rid == 'prb3_review':
        if not refs['skill_arts_prb3_submit_start']:
            return ('err', 'PRB3 review scheduled before PRB3 submit (internal error).')
        ideal_review = add_working_days_utc(refs['skill_arts_prb3_submit_start'], skill_arts_tier_span - 1, holidays)
        resolved = resolve_prb_anchor_day(ideal_review, holidays, phase_allow)
        start = str(resolved['iso'])
        end = add_working_days_utc(start, effective - 1, holidays)
        steps_out.append(
            linear_step_payload(
                str(row['label']),
                start,
                end,
                note_out,
                linear_prb_step_allow_non_working(
                    phase_allow, start, holidays, bool(resolved['needsAllowNonWorkingFlag'])
                ),
            )
        )
        return ('placed', add_calendar_days_utc(end, 1))
    return ('skip', None)


def skill_arts_try_place_tiered_development_dev_gap_row(
    timing_profile: str,
    allow: bool,
    row: Mapping[str, Any],
    holidays: HolidaySet,
    allow_map: dict[str, bool] | None,
    effective: int,
    note_out: str,
    skill_arts_tier_span: int,
    refs: SkillArtsPrbRefs,
    steps_out: list[dict[str, Any]],
) -> tuple[Literal['placed'], str] | tuple[Literal['skip'], None] | tuple[Literal['err'], str]:
    rid = str(row['id'])
    dev_prb_row = rid in ('development_prb1', 'development_prb2', 'development_prb3')
    use_skillarts_tiered_dev_gap = uses_skillarts_tiered_prb_cadence(timing_profile) and not allow and dev_prb_row
    if not use_skillarts_tiered_dev_gap:
        return ('skip', None)

    submit_sa = (
        refs['skill_arts_prb1_submit_start']
        if rid == 'development_prb1'
        else refs['skill_arts_prb2_submit_start']
        if rid == 'development_prb2'
        else refs['skill_arts_prb3_submit_start']
    )
    submit_end_sa = (
        refs['skill_arts_prb1_submit_end']
        if rid == 'development_prb1'
        else refs['skill_arts_prb2_submit_end']
        if rid == 'development_prb2'
        else refs['skill_arts_prb3_submit_end']
    )
    review_phase_sa = (
        'prb1_review' if rid == 'development_prb1' else 'prb2_review' if rid == 'development_prb2' else 'prb3_review'
    )
    if submit_sa is None or submit_end_sa is None:
        return ('err', 'PRB development gap missing submit anchor (internal error).')
    ideal_review_sa = add_working_days_utc(submit_sa, skill_arts_tier_span - 1, holidays)
    review_allow_sa = linear_allow_non_working_for(review_phase_sa, allow_map)
    review_resolved_sa = resolve_prb_anchor_day(ideal_review_sa, holidays, review_allow_sa)
    gap_start = next_working_day(add_calendar_days_utc(submit_end_sa, 1), holidays)
    gap_end = previous_working_day_on_or_before(
        add_calendar_days_utc(str(review_resolved_sa['iso']), -1), holidays
    )
    if parse_iso_date_utc(gap_start) > parse_iso_date_utc(gap_end):
        gap_start = gap_end
    start = gap_start
    end = add_working_days_utc(start, effective - 1, holidays)
    steps_out.append(linear_step_payload(str(row['label']), start, end, note_out, False))
    return ('placed', add_calendar_days_utc(end, 1))

File: apps/web-dashboard/lib/scenarioPlanner/linear/skillarts_strategy.ts
/**
 * SkillArts tiered PRB cadence (Thursday submit anchor, page-tier working days submit→review).
 */

import type { HalTimelineStep } from '../../halScenario';
import { addCalendarDaysUTC, parseIsoDateUTC } from '../dateCalendar';
import {
  firstWorkingThursdayOnOrAfter,
  resolvePrbAnchorDay,
} from '../prbWeekdayAnchors';
import type { ScenarioTactic } from '../tactics';
import { usesSkillArtsTieredPrbCadence } from '../timingProfiles';
import {
  addWorkingDaysUTC,
  inclusiveWorkingDaySpan,
  nextWorkingDay,
  previousWorkingDayOnOrBefore,
  type HolidaySet,
} from '../workingDays';
import type { ScenarioStepDef } from './types';
import { linearAllowNonWorkingFor, linearClampDays, linearPrbStepAllowNonWorking, linearStepPayload } from './scenario_linear_shared';

const SA_PRB_IDS = new Set([
  'submit_prb1',
  'prb1_review',
  'submit_prb2',
  'prb2_review',
  'submit_prb3',
  'prb3_review',
]);

export type SkillArtsPrbRefs = {
  skillArtsPrb1SubmitStart: string | null;
  skillArtsPrb1SubmitEnd: string | null;
  skillArtsPrb2SubmitStart: string | null;
  skillArtsPrb2SubmitEnd: string | null;
  skillArtsPrb3SubmitStart: string | null;
  skillArtsPrb3SubmitEnd: string | null;
};

export function skillArtsApplyPinnedPrefixPrbState(
  timingProfile: ScenarioTactic,
  allow: boolean,
  row: ScenarioStepDef,
  pin: HalTimelineStep,
  refs: SkillArtsPrbRefs,
): void {
  if (!usesSkillArtsTieredPrbCadence(timingProfile) || allow || !SA_PRB_IDS.has(row.id)) return;
  if (row.id === 'submit_prb1') {
    refs.skillArtsPrb1SubmitStart = pin.start_date;
    refs.skillArtsPrb1SubmitEnd = pin.end_date;
  } else if (row.id === 'submit_prb2') {
    refs.skillArtsPrb2SubmitStart = pin.start_date;
    refs.skillArtsPrb2SubmitEnd = pin.end_date;
  } else if (row.id === 'submit_prb3') {
    refs.skillArtsPrb3SubmitStart = pin.start_date;
    refs.skillArtsPrb3SubmitEnd = pin.end_date;
  }
}

export function skillArtsAdjustEffectiveForDevelopmentDevGap(
  timingProfile: ScenarioTactic,
  allow: boolean,
  row: ScenarioStepDef,
  holidays: HolidaySet,
  allowMap: Readonly<Record<string, boolean>> | undefined,
  modSum: number,
  effective: number,
  skillArtsTierSpan: number,
  refs: SkillArtsPrbRefs,
): number | { ok: false; error: string } {
  const devPrbRow =
    row.id === 'development_prb1' || row.id === 'development_prb2' || row.id === 'development_prb3';
  const useSkillArtsTieredDevGap =
    usesSkillArtsTieredPrbCadence(timingProfile) && !allow && devPrbRow;
  if (!useSkillArtsTieredDevGap) return effective;

  let submitAnchor: string | null = null;
  let submitEnd: string | null = null;
  let reviewPhaseId: 'prb1_review' | 'prb2_review' | 'prb3_review';
  let errEarly: string | null = null;
  if (row.id === 'development_prb1') {
    submitAnchor = refs.skillArtsPrb1SubmitStart;
    submitEnd = refs.skillArtsPrb1SubmitEnd;
    reviewPhaseId = 'prb1_review';
    errEarly = 'PRB1 development scheduled before PRB1 submit (internal error).';
  } else if (row.id === 'development_prb2') {
    submitAnchor = refs.skillArtsPrb2SubmitStart;
    submitEnd = refs.skillArtsPrb2SubmitEnd;
    reviewPhaseId = 'prb2_review';
    errEarly = 'PRB2 development scheduled before PRB2 submit (internal error).';
  } else {
    submitAnchor = refs.skillArtsPrb3SubmitStart;
    submitEnd = refs.skillArtsPrb3SubmitEnd;
    reviewPhaseId = 'prb3_review';
    errEarly = 'PRB3 development scheduled before PRB3 submit (internal error).';
  }
  if (!submitAnchor || !submitEnd) {
    return { ok: false, error: errEarly! };
  }
  const idealReview = addWorkingDaysUTC(submitAnchor, skillArtsTierSpan - 1, holidays);
  const reviewAllow = linearAllowNonWorkingFor(reviewPhaseId, allowMap);
  const reviewResolved = resolvePrbAnchorDay(idealReview, holidays, reviewAllow);
  let gapStart = nextWorkingDay(addCalendarDaysUTC(submitEnd, 1), holidays);
  const gapEnd = previousWorkingDayOnOrBefore(addCalendarDaysUTC(reviewResolved.iso, -1), holidays);
  if (parseIsoDateUTC(gapStart).getTime() > parseIsoDateUTC(gapEnd).getTime()) {
    gapStart = gapEnd;
  }
  const rawSpan = Math.max(1, inclusiveWorkingDaySpan(gapStart, gapEnd, holidays));
  const bumped = linearClampDays(rawSpan + modSum, row);
  return Math.min(bumped, rawSpan);
}

export function skillArtsTryPlaceTieredPrbRow(
  timingProfile: ScenarioTactic,
  allow: boolean,
  row: ScenarioStepDef,
  cursorCal: string,
  holidays: HolidaySet,
  allowMap: Readonly<Record<string, boolean>> | undefined,
  effective: number,
  noteOut: string,
  skillArtsTierSpan: number,
  refs: SkillArtsPrbRefs,
  steps: HalTimelineStep[],
): { placed: true; nextCursorCal: string } | { placed: false } | { ok: false; error: string } {
  const useSkillArtsTieredCadence =
    usesSkillArtsTieredPrbCadence(timingProfile) &&
    !allow &&
    SA_PRB_IDS.has(row.id);
  if (!useSkillArtsTieredCadence) return { placed: false };

  const phaseAllow = linearAllowNonWorkingFor(row.id, allowMap);

  if (row.id === 'submit_prb1') {
    const minStart = nextWorkingDay(cursorCal, holidays);
    const merged = minStart;
    const thuRaw = firstWorkingThursdayOnOrAfter(merged, holidays);
    const resolved = resolvePrbAnchorDay(thuRaw, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    refs.skillArtsPrb1SubmitStart = start;
    refs.skillArtsPrb1SubmitEnd = end;
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  if (row.id === 'prb1_review') {
    if (!refs.skillArtsPrb1SubmitStart) {
      return { ok: false, error: 'PRB1 review scheduled before PRB1 submit (internal error).' };
    }
    const idealReview = addWorkingDaysUTC(refs.skillArtsPrb1SubmitStart, skillArtsTierSpan - 1, holidays);
    const resolved = resolvePrbAnchorDay(idealReview, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  if (row.id === 'submit_prb2') {
    const minStart = nextWorkingDay(cursorCal, holidays);
    const merged = minStart;
    const thuRaw = firstWorkingThursdayOnOrAfter(merged, holidays);
    const resolved = resolvePrbAnchorDay(thuRaw, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    refs.skillArtsPrb2SubmitStart = start;
    refs.skillArtsPrb2SubmitEnd = end;
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  if (row.id === 'prb2_review') {
    if (!refs.skillArtsPrb2SubmitStart) {
      return { ok: false, error: 'PRB2 review scheduled before PRB2 submit (internal error).' };
    }
    const idealReview = addWorkingDaysUTC(refs.skillArtsPrb2SubmitStart, skillArtsTierSpan - 1, holidays);
    const resolved = resolvePrbAnchorDay(idealReview, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  if (row.id === 'submit_prb3') {
    const minStart = nextWorkingDay(cursorCal, holidays);
    const merged = minStart;
    const thuRaw = firstWorkingThursdayOnOrAfter(merged, holidays);
    const resolved = resolvePrbAnchorDay(thuRaw, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    refs.skillArtsPrb3SubmitStart = start;
    refs.skillArtsPrb3SubmitEnd = end;
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  if (row.id === 'prb3_review') {
    if (!refs.skillArtsPrb3SubmitStart) {
      return { ok: false, error: 'PRB3 review scheduled before PRB3 submit (internal error).' };
    }
    const idealReview = addWorkingDaysUTC(refs.skillArtsPrb3SubmitStart, skillArtsTierSpan - 1, holidays);
    const resolved = resolvePrbAnchorDay(idealReview, holidays, phaseAllow);
    const start = resolved.iso;
    const end = addWorkingDaysUTC(start, effective - 1, holidays);
    steps.push(
      linearStepPayload(
        row.label,
        start,
        end,
        noteOut,
        linearPrbStepAllowNonWorking(phaseAllow, start, holidays, resolved.needsAllowNonWorkingFlag),
      ),
    );
    return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
  }
  return { placed: false };
}

export function skillArtsTryPlaceTieredDevelopmentDevGapRow(
  timingProfile: ScenarioTactic,
  allow: boolean,
  row: ScenarioStepDef,
  holidays: HolidaySet,
  allowMap: Readonly<Record<string, boolean>> | undefined,
  effective: number,
  noteOut: string,
  skillArtsTierSpan: number,
  refs: SkillArtsPrbRefs,
  steps: HalTimelineStep[],
): { placed: true; nextCursorCal: string } | { placed: false } {
  const devPrbRow =
    row.id === 'development_prb1' || row.id === 'development_prb2' || row.id === 'development_prb3';
  const useSkillArtsTieredDevGap =
    usesSkillArtsTieredPrbCadence(timingProfile) && !allow && devPrbRow;
  if (!useSkillArtsTieredDevGap) return { placed: false };

  const submitAnchor =
    row.id === 'development_prb1'
      ? refs.skillArtsPrb1SubmitStart!
      : row.id === 'development_prb2'
        ? refs.skillArtsPrb2SubmitStart!
        : refs.skillArtsPrb3SubmitStart!;
  const submitEnd =
    row.id === 'development_prb1'
      ? refs.skillArtsPrb1SubmitEnd!
      : row.id === 'development_prb2'
        ? refs.skillArtsPrb2SubmitEnd!
        : refs.skillArtsPrb3SubmitEnd!;
  const reviewPhaseId =
    row.id === 'development_prb1' ? 'prb1_review' : row.id === 'development_prb2' ? 'prb2_review' : 'prb3_review';
  const idealReview = addWorkingDaysUTC(submitAnchor, skillArtsTierSpan - 1, holidays);
  const reviewResolved = resolvePrbAnchorDay(
    idealReview,
    holidays,
    linearAllowNonWorkingFor(reviewPhaseId, allowMap),
  );
  let gapStart = nextWorkingDay(addCalendarDaysUTC(submitEnd, 1), holidays);
  const gapEnd = previousWorkingDayOnOrBefore(addCalendarDaysUTC(reviewResolved.iso, -1), holidays);
  if (parseIsoDateUTC(gapStart).getTime() > parseIsoDateUTC(gapEnd).getTime()) {
    gapStart = gapEnd;
  }
  const start = gapStart;
  const end = addWorkingDaysUTC(start, effective - 1, holidays);
  steps.push(linearStepPayload(row.label, start, end, noteOut, false));
  return { placed: true, nextCursorCal: addCalendarDaysUTC(end, 1) };
}